### PR TITLE
[4.0] Contact - nullable datetime fields

### PR DESCRIPTION
**Pull request by alikon**
Pull Request for Issue #16788 ....same as  #21901 for modules

### Summary of Changes

some datetime fields should be nullable as default

- publish_up
- publish_down
- checked_out_time

### Testing Instructions
test  com_contacts 

### Expected result
contacts should works as before

### CODE REVIEW

File: administrator/components/com_contact/Table/ContactTable.php [hunk 1]
```diff
@@ -27,6 +27,15 @@
  */
 class ContactTable extends Table
 {
+
```
Comment (zero-24) [inline administrator/components/com_contact/Table/ContactTable.php]: please remove the clear line

File: components/com_contact/Model/CategoryModel.php [hunk 2]
```diff
@@ -179,9 +179,9 @@ protected function getListQuery()
 		$nowDate = $db->quote(Factory::getDate()->toSql());
 
 		if ($this->getState('filter.publish_date'))
-		{
-			$query->where('(a.publish_up = ' . $nullDate . ' OR a.publish_up <= ' . $nowDate . ')')
-				->where('(a.publish_down = ' . $nullDate . ' OR a.publish_down >= ' . $nowDate . ')');
+		{
```
Comment (zero-24) [inline components/com_contact/Model/CategoryModel.php]: to much spaces

**Review by zero-24 — DISMISSED** [review]

File: administrator/components/com_contact/Table/ContactTable.php (hunk 1 above)
Comment (alikon) [reply administrator/components/com_contact/Table/ContactTable.php]: done

**Comment by wilsonge** [comment]
Open question (can see it both ways). Should modified date by nullable or should enforce it being date time (modified date is the created date on creation).

**Comment by alikon** [comment]
yes, to me too, we need to take a decision and make it consistently across the cms

**Comment by twister65** [comment]
I have tested this item :white_check_mark: successfully on 44b90c5a7778dba24f249b8dba0d4f4ac6db08d7<br /><br />with MySQL and PostgreSQL.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/24675">issues.joomla.org/tracker/joomla-cms/24675</a>.</sub>

**Comment by wilsonge** [comment]
Unless anyone feels strongly I'd rather have this set all the time. So modified date is always set (to the created date time when the article is created)

**Comment by alikon** [comment]
ok `modified` field `is not null ` now

p.s.
i hope prepared statement should be room for a new pr

File: administrator/components/com_contact/Table/ContactTable.php [hunk 3]
```diff
@@ -253,9 +249,20 @@ public function check()
 			$this->metadata = '{}';
 		}
 
-		if (empty($this->modified))
+		// Set publish_up, publish_down, modified to null if not set
```
Comment (Quy) [inline administrator/components/com_contact/Table/ContactTable.php]: Remove `modified`

Comment (Quy) [inline administrator/components/com_contact/Table/ContactTable.php]:
```suggestion
		// Set publish_up, publish_down to null if not set
```

**Comment by wilsonge** [comment]
There's no queries here that require prepared statements. This looks fine. Just needs some tests creating and updating contacts

**Comment by alikon** [comment]
are you sure ? 
just for example components/com_contact/Model/CategoryModel.php is full of places where we can use prepared statements https://github.com/joomla/joomla-cms/blob/c5ae1414988e75d38a55493ed595e272d17a0317/components/com_contact/Model/CategoryModel.php#L193

https://github.com/joomla/joomla-cms/blob/c5ae1414988e75d38a55493ed595e272d17a0317/components/com_contact/Model/CategoryModel.php#L170

**Comment by Quy** [comment]
I have tested this item :white_check_mark: successfully on 456218b31b332951922c4ed89d14d82cda8ce975<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/24675">issues.joomla.org/tracker/joomla-cms/24675</a>.</sub>

**Comment by Quy** [comment]
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/24675">issues.joomla.org/tracker/joomla-cms/24675</a>.</sub>

**Comment by wilsonge** [comment]
Thanks!

**Pull request merged**